[Stream] Improve the function signature verifier for CmdDispatchOp

qedawkins opened the pull request:
This adds additional verification to the exported function within
a dispatch that the binding counts and uniform operand types are
consistent with the dispatch op.

qedawkins committed:
[Stream] Only fuse bindings when the function represents them all

benvanik commented on compiler/src/iree/compiler/Dialect/Stream/Transforms/FuseDispatchBindings.cpp:
```diff
@@ -322,6 +322,25 @@ fuseDispatchBindings(IREE::Stream::ExecutableOp executableOp,
+  // bindings, in which case we can no longer reliably fuse them. We could try
+  // to be clever and fuse a subset of bindings, but this typically only happens
+  // when the user is trying to do something custom.
+  if (numBindingsOnFunc != bindingCount)
```
this seems a bit too complex - are there any cases where we'd have non-zero partial bindings?
(would rather keep this simple: if there are bindings we optimize them, if there aren't we don't)

qedawkins replied:
The only case I can think of would be if a user decided to write it like that, so maybe not worth ever supporting.

benvanik replied:
it'd also be good to understand what you're doing here - we really shouldn't be use stream ops without bindings and instead generate hal ops directly - the only reason to use stream ops is to get the stuff stream dialect can do

qedawkins replied:
I was doing this: https://gist.github.com/qedawkins/1f344eb7551d063e360ce9d3a85ab0a7

Using a `flow.executable`. My basic thought process was if I wanted to use mostly the same input IR for CPU and SPIRV, I should go through Flow ops instead of having to create variants with specific targets (even though in the example I put SPIR-V codegen configs).

If this IR shouldn't be allowed, then we could strengthen the verifier of `flow.executable`, else it will turn into stream IR and then crash here without this change.

benvanik replied:
yeah, that's a misuse - flow/stream are expecting flow/stream stuff, not hal - you don't need to specify hal bindings and such - just use the appropriate flow ops (`flow.dispatch.workgroup.id`, etc).

we can't verify that users don't put in random things - we allow custom ops and such - from the flow/stream perspective the hal is just some random other dialect

qedawkins replied:
ah `flow.dispatch.workgroup.id`, let me see how that works.

benvanik replied:
yeah give that a go and if there's stuff missing we can add additional plumbing/routes through - I'm pretty sure we don't want to support this kind of usage (using hal ops with flow executables) so I'm not sure we want to land this extra logic (there shouldn't be cases where there are flow/stream executables with bodies that we can't touch - lots of passes aren't expecting that)

qedawkins replied:
Makes sense, I guess it was kind of hard to tell that this was unexpected usage, at least when I was comparing against external objects on `flow.executables` (the precursor to `hal.dispatch.extern`). It seemed like we were already doing this for those cases, and doing the same for something with an explicit body that we expect some kind of codegen+target to handle didn't seem very different. Also this was the only pass to crash :P

(side note: really wish Github didn't require a reload to get messages in conversations like this. Makes it hard to track responses)

qedawkins replied:
nice, that worked. I can either close this or change it to signal a pass failure with some kind of error message when the function signature doesn't include enough bindings (in case a user tries to do something like this in the future).

benvanik replied:
verifying that binding count matches would be awesome - AsyncDispatchOp/CmdDispatchOp have a TODO for "verify that the target function has matching operands" that'd be good to resolve - it currently just checks the workload -> workgroup count region

qedawkins replied:
I just did it for CmdDispatchOp for now, but it might be easier to just do both at the same time hmmm

qedawkins committed:
Add test and simplify comment

qedawkins committed:
Revert "Add test and simplify comment"

This reverts commit be72687f8651f1ae7ec21e47b7f4f88799d06050.

qedawkins committed:
Revert "[Stream] Only fuse bindings when the function represents them all"

This reverts commit 404d2f3e056e546213041f3c8ffde29607dfaaaf.

qedawkins committed:
[Stream] Improve the function signature verifier for CmdDispatchOp

This adds additional verification to the exported function within
a dispatch that the binding counts and uniform operand types are
consistent with the dispatch op.

qedawkins committed:
Fix invalid.mlir run command

qedawkins committed:
Require strict type matching and update complex test

benvanik commented on compiler/src/iree/compiler/Dialect/Stream/IR/test/invalid.mlir:
```diff
@@ -0,0 +1,27 @@
+// RUN: iree-opt --split-input-file %s --verify-diagnostics
```
these can be tested in executable_ops.mlir alongside the other ones, right?

qedawkins replied:
yes, I'm just used to `invalid.mlir` and didn't see one. Moved now.

qedawkins committed:
Remove invalid.mlir
